src/main/java/com/rockset/jdbc/RocksetResultSet.java

```diff
@@ -1579,22 +1595,23 @@ private static List<Column> getColumns(QueryResponse response) throws SQLExcepti
           Object onedoc = response.getResults().get(i);
           JsonNode docRootNode = mapper.readTree(mapper.writeValueAsString(onedoc));
 
-          Iterator<Map.Entry<String, JsonNode>> fields = docRootNode.fields();
-          while (fields.hasNext()) {
-            Map.Entry<String, JsonNode> field = fields.next();
-            String fieldName = field.getKey();
+          Iterator<Map.Entry<String, JsonNode>> rootEntries = docRootNode.fields();
+          Map<String, JsonNode> root = new LinkedHashMap<>();
+          while (rootEntries.hasNext()) {
+            Map.Entry<String, JsonNode> field = rootEntries.next();
+            root.put(field.getKey(), field.getValue());
+          }
+          // If we know the desired field ordering through explicit projection, use that ordering.
+          // Otherwise, just iterate over in arbitrary ordering
+          Collection<String> fields = orderedFields.isEmpty() ? root.keySet() : orderedFields;
+          for (String fieldName: fields) {
             if (fieldNames.contains(fieldName)) {
               // This fieldname was already found to have a non-null type
               // in a previous record.
               continue;
             }
-            JsonNode value = field.getValue();
+            JsonNode value = root.get(fieldName);
             Column.ColumnTypes type = Column.ColumnTypes.fromValue(value.getNodeType().toString());
-            // Skip over the fields with null type unless all values for that
-            // field are null
-            if (type.equals(Column.ColumnTypes.NULL) && i != response.getResults().size() - 1) {
-              continue;
-            }
```

Comment:
We may need something more subtle here. If the first row has null but second has a string, the column type should be string. I think this code will infer null now.